### PR TITLE
[java] UseArraysAsList fixes

### LINKED ISSUE #3379: [java] UseArraysAsList must ignore primitive arrays
**Affects PMD Version:** 
6.36.0

**Rule:**
[UseArraysAsList](https://pmd.github.io/pmd-6.36.0/pmd_rules_java_performance.html#usearraysaslist)

**Description:**
Arrays of primitive data types must not trigger this rule.
This is especially problematic since replacing the code as suggested would introduce a bug, i.e. an Array with a single entry instead of an Array of values converted from primitive type to Object type (e.g. `short` into `Short`).

The following arrays should not trigger the rule:

- `byte[].class`
- `short[].class`
- `int[].class`
- `long[].class`
- `float[].class`
- `double[].class`
- `boolean[].class`
- `char[].class`

(I hope I didn't miss any.)

**Code Sample demonstrating the issue:**

```java
short[] array = (short[]) value;
ArrayList<Short> arrayList = new ArrayList<>(array.length);
for(short v:array) {
    arrayList.add(v);
}

// can't use this:
var list2 = Arrays.asList(array); // type of list2: List<short[]> and not List<Short>
```

**Expected outcome:**

PMD reports a violation at line `arrayList.add(v);` but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

### LINKED ISSUE #3965: [java] UseArraysAsList false positive with non-trivial loops
**Affects PMD Version:** 6.45.0

**Rule:** [UseArraysAsList](https://pmd.github.io/latest/pmd_rules_java_performance.html#usearraysaslist)

**Description:** The rule should only report simple, trivial loops, that do nothing more than converting an array into a list.

**Code Sample demonstrating the issue:**

```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Test {
    public void foo(Integer[] ints) {
        List<Integer> l = new ArrayList<>(100);
        for (int i = 0; i < 100; i++) {
            switch (lookup(ints[i])) {
            case 1: l.add(ints[i]); break; // line 10 - false positive
            case 2: l.addAll(getInts(i)); break;
            }
        }

        List<Integer> anotherList = new ArrayList<>();
        for (int i : ints) {
            switch (lookup(i)) {
            case 1: anotherList.add(i); break; // line 18 - false positive
            case 2: anotherList.addAll(getInts(i)); break;
            }
        }
    }

    int lookup(int a) {
        return a;
    }

    List<Integer> getInts(int a) {
        return Arrays.asList(a);
    }
}
```

**Expected outcome:**

PMD reports a violation at lines 10 and 18, but that's wrong. That's a false positive.

### PR DESCRIPTION
**Pull request by adangel**
- Fixes #3379 
- Fixes #3965 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/performance.xml [hunk 1]
```diff
@@ -968,7 +968,14 @@ You must use `new ArrayList<>(Arrays.asList(...))` if that is inconvenient for y
 <![CDATA[
 //Statement/ForStatement
     [not(.//IfStatement)]
+    [count(Statement//StatementExpression)=1]
     [@Foreach = true() or ForInit//VariableInitializer//Literal[@IntLiteral= true() and @Image='0']]
+    (: skip primitive types :)
+    [@Foreach = true() and Expression[not(pmd-java:typeIs('boolean[]') or pmd-java:typeIs('char[]') or pmd-java:typeIs('byte[]') or pmd-java:typeIs('short[]') or pmd-java:typeIs('int[]') or pmd-java:typeIs('long[]') or pmd-java:typeIs('float[]') or pmd-java:typeIs('double[]'))]
+        or Expression/RelationalExpression/PrimaryExpression/PrimaryPrefix/Name[substring-before(@Image, '.length') =
+           ancestor::MethodDeclaration[1]//VariableDeclaratorId[not(pmd-java:typeIs('boolean[]') or pmd-java:typeIs('char[]') or pmd-java:typeIs('byte[]') or pmd-java:typeIs('short[]') or pmd-java:typeIs('int[]') or pmd-java:typeIs('long[]') or pmd-java:typeIs('float[]') or pmd-java:typeIs('double[]'))]
+           /@Name]
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/performance.xml:977]: That should work too, right? We might get some FNs if type resolution fails us, but I think this is ok...

```suggestion
    [@Foreach = true() and Expression[pmd-java:typeIs('java.lang.Object[]')]
        or Expression/RelationalExpression/PrimaryExpression/PrimaryPrefix/Name[substring-before(@Image, '.length') =
           ancestor::MethodDeclaration[1]//VariableDeclaratorId[pmd-java:typeIs('java.lang.Object[]')]
           /@Name]
```

Comment (adangel) [reply pmd-java/src/main/resources/category/java/performance.xml:977]: Let's try :)

**Review by oowekyala — APPROVED** [review]
LGTM, thanks!

**Pull request merged**